### ISSUE TITLE
Pythonista environment variables

### ISSUE THREAD

**Issue by zrzka** [omz/Pythonista-Issues#363]
I just started adding some tests to Black Mamba. Did add two very simple tests and now figuring out how to properly integrate it with Travis CI.

To make it work on Travis CI (basically anywhere outside of Pythonista), I do use `mock` and `MagicMock()`. Check [tests/__init__.py](https://github.com/zrzka/blackmamba/blob/master/tests/__init__.py).

Would be nice if Pythonista can provide two environment variables like:

* `PYTHONISTA=true`
* `PYTHONISTA_VERSION=3...`

Like Travis provides `CI=true`, etc. Or is there any other way how to do it more effectively?

**Comment by TeamSpen210** [comment]
You could also check if `sys.platform == "ios"`, and check the normal `sys.version_info.major` for 2/3 detection.

**Comment by zrzka** [comment]
Yup, but it doesn't say it's running under Pythonista (kinda yes, but ...). The `sys.version_info` reports Python version. I meant Pythonista version.

**Comment by TeamSpen210** [comment]
Perhaps Pythonista should set `sys.implementation` differently, although it still is CPython. In that case it could add `_pythonista_version` for example set to whatever (That's the [purpose](https://www.python.org/dev/peps/pep-0421) of the attributes).

**Comment by cclauss** [comment]
Ten lines or less has Pythonista_versions.py https://github.com/cclauss/Ten-lines-or-less/tree/master

**Comment by zrzka** [comment]
@cclauss I can read Pythonista version as well, but that's not the point. Point is easy detection if the script is running under the Pythonista or not. Without any things like checking if plist file exists, if plistinfo package is installed, ... Something simple like env variable.

**Comment by cclauss** [comment]
`sys.platform == "ios"`

**Comment by zrzka** [comment]
That's what I have now (environ) https://github.com/zrzka/blackmamba/blob/master/tests/__init__.py But did write this issue because I don't like it, same for sys.platform :)

**Comment by cclauss** [comment]
Ahhh...  You don't like it.  Now I understand.

**Comment by dgelessus** [comment]
Different suggestion - if all you want to do is mock the Pythonista modules if they don't exist, you can `try` the imports, `catch ImportError` and do the mocking only if the imports failed.

Back on topic, IMHO environment variables are not a great solution. The reason why Travis uses envvars is because they are easily accessible from any decent programming language. Pythonista only has Python, so IMHO such a feature should be implemented in native Python. For example there could be a `pythonista` module with a `__version__` attribute.

**Comment by zrzka** [comment]
@dgelessus as always, elegant solution with `pythonista` and version.

Re mock - it's just temporary what you saw. I'm still thinking about the right solution. Black Mamba manipulates with user files (toggle comments, analyzer, ...) and I would like to cover these things with tests. Tests should be runnable under the Pythonista (where `editor` exists) and on Travis CI (where `editor` doesn't exist). Travis solution is quite easy with mock. But running these tests under Pythonista will be more complicated, because I would like to test `editor...` calls as well and do not want to manipulate with real files, should be kind of mock as well. Thinking of `@patch`, ... But didn't play with it & Pythonista yet. Anyway, it's not a good topic for this issue :) I'll probably start new topic in the forum if I'll be lost.

**Comment by omz** [comment]
A `pythonista` module sounds good, but what would be in there, except for `__vesion__`? Would it perhaps make more sense to add a `pythonista_version` attribute to the `editor` module?

**Comment by cclauss** [comment]
Pythonista version 3.1.1 (311008) running Python 3.6.1 on iOS 10.3.3 on a 64-bit iPad5,4 with a screen size of (1024 x 768) * 2

**Comment by zrzka** [comment]
I like `pythonista` module (compared to the `editor.pythonista_version`) more, so, I can do:

```python
try
    import pythonista
except ...
    # Okay, I'm outside of Pythonista
```

It's much clearer what's going on for readers compared to:

```python
try:
    import editor
except ...
    # Outside of Pythonista? What is editor? Should I comment it?
```

Also I would expect `'3.1.1'` in the `__version__` (PEP396). What @cclauss suggests can be in any other attribute.

**Comment by omz** [comment]
@zrzka Makes sense. Maybe the `pythonista` module should be an alias for `editor`?

**Comment by zrzka** [comment]
@omz I'm not sure if understand what do you mean? `import editor as pythonista` instead of `import pythonista`?

**Comment by omz** [comment]
@zrzka I actually meant that `pythonista.py` could essentially be

```
from editor import *
__version__ = '3.2'
```

**Comment by dgelessus** [comment]
Ideas for things that could be in the `pythonista` module:

* Pythonista version (duh)
* Pythonista build number
* Beta or not (boolean flag)
* Supported Python versions (all of them, not just the current one)

Though I also like the idea of `editor.pythonista_version`. Perhaps Editorial's editor could have an `editor.editorial_version`, then we'd have an easy way to tell apart Pythonista and Editorial.

**Comment by cclauss** [comment]
For me ,`editor.__version__` would be the version of the editor module, not the version of Pythonista.  Perhaps `editor.__pythonista_version__`?

**Comment by dgelessus** [comment]
Good point. On second thought, I'm not sure if the dunders should be there at all. Maybe it should be `pythonista.version` and/or `editor.pythonista_version`? Dunder attributes on modules are usually for metadata about the module (`__version__`, `__author__`, `__license__`), and the Pythonista version isn't really metadata of the `pythonista` or `editor` module.

**Comment by zrzka** [comment]
I tend to agree with `pythonista.version` instead of `pythonista.__version__`. I think it will be much more clear what it contains in case other stuff like already mentioned `.beta`, `.build_number`, ... will be added as well.

**Comment by TeamSpen210** [comment]
See [PEP-421](https://www.python.org/dev/peps/pep-0421), this is what `sys.implementation` is supposed to be for - describing the version details for the implementation of the language, not the version of it.

**Comment by cclauss** [comment]
@TeamSpen210 from Pythonista beta today:
```python
>>> import sys
>>> sys.implementation
namespace(_multiarch='darwin', cache_tag='cpython-36', hexversion=50725360, name='cpython', version=sys.version_info(major=3, minor=6, micro=1, releaselevel='final', serial=0))
```

This is correct but does not help you know if you are in Editorial, Pythonista, or Mac OSX as all are `Darwin`.

**Comment by TeamSpen210** [comment]
The PEP indicates that arbitrary `_attributes` can be added, so a `_pythonista_version` attribute could be added for example.

**Comment by omz** [comment]
So far, I like `sys.pythonista_version` best (and maybe `_pythonista_version_info` as a numeric tuple).

**Comment by cclauss** [comment]
I like @TeamSpen210 approach better:
```python
def is_pythonista():
    return _pythonista_version in sys.implementation

# vs.
def is_pythonista():
    try:
        sys._pythonista_version
        return True
    except AttributeError
        return False
```

You could play dir() or has_attr tricks but is is not fluid.

**Comment by dgelessus** [comment]
`in` doesn't work with `sys.implementation`:

```python
>>> sys.implementation
Out[23] = namespace(_multiarch='darwin', cache_tag='cpython-36', hexversion=50725360, name='cpython', version=sys.version_info(major=3, minor=6, micro=1, releaselevel='final', serial=0))
>>> "version" in sys.implementation
Traceback (most recent call last):
	File <string>, line 1, in <module>:
		# Source code unavailable

builtins.TypeError: argument of type 'types.SimpleNamespace' is not iterable
```

Also I think `hasattr(sys, "pythonista_version")` is quite short already.

**Comment by zrzka** [comment]
Here're examples what / how do use this info in Black Mamba - [system.py](https://github.com/zrzka/blackmamba/blob/v1.3.2/blackmamba/system.py).
